language/move-vm/runtime/src/loader.rs

```diff
@@ -0,0 +1,1378 @@
+            None => Ok(Arc::clone(self.modules.insert(id, module))),
+        }
+    }
+
+    fn function_at(&self, idx: usize) -> Arc<Function> {
+        self.functions[idx].clone()
+    }
+
+    fn struct_at(&self, idx: usize) -> Arc<StructType> {
+        Arc::clone(&self.structs[idx])
+    }
+
+    fn publish_module(&mut self, module: &VerifiedModule) -> VMResult<()> {
+        let starting_idx = self.structs.len();
+        for (idx, struct_def) in module.struct_defs().iter().enumerate() {
+            let st = self.load_type(module, struct_def, StructDefinitionIndex(idx as u16))?;
+            self.structs.push(Arc::new(st));
+        }
+        self.load_fields(module, starting_idx)?;
+
+        for func in module.function_defs() {
+            let function = self.load_function(module, func)?;
+            self.functions.push(Arc::new(function));
+        }
+        Ok(())
+    }
+
+    fn load_type(
+        &self,
+        module: &VerifiedModule,
+        struct_def: &StructDefinition,
+        idx: StructDefinitionIndex,
+    ) -> VMResult<StructType> {
+        let struct_handle = module.struct_handle_at(struct_def.struct_handle);
+        let is_resource = struct_handle.is_nominal_resource;
+        let name = module.identifier_at(struct_handle.name).to_owned();
+        let type_parameters = struct_handle.type_parameters.clone();
+        let module = module.self_id();
+        Ok(StructType {
+            fields: vec![],
+            is_resource,
+            type_parameters,
+            name,
+            module,
+            struct_def: idx,
+        })
+    }
+
+    fn load_fields(&mut self, module: &VerifiedModule, starting_idx: usize) -> VMResult<()> {
+        let mut field_types = vec![];
+        for struct_def in module.struct_defs() {
+            let fields = match &struct_def.field_information {
+                StructFieldInformation::Native => unreachable!("native structs have been removed"),
+                StructFieldInformation::Declared(fields) => fields,
+            };
+
+            let mut field_tys = vec![];
+            for field in fields {
+                let ty = self.make_type(module.as_inner(), &field.signature.0)?;
+                assume!(field_tys.len() < usize::max_value());
+                field_tys.push(ty);
+            }
+
+            field_types.push(field_tys);
+        }
+        for (fields, arc_struct_type) in field_types
+            .into_iter()
+            .zip(&mut self.structs[starting_idx..])
+        {
+            match Arc::get_mut(arc_struct_type) {
+                None => {
+                    return Err(VMStatus::new(StatusCode::INVALID_CODE_CACHE)
+                        .with_message("Arc Type should not have any reference".to_string()))
+                }
+                Some(struct_type) => struct_type.fields = fields,
+            }
+        }
+        Ok(())
+    }
+
+    fn make_type(&self, binary: &dyn ModuleAccess, tok: &SignatureToken) -> VMResult<Type> {
```

Comment:
Can we use "impl ModuleAccess" or `make_type<M: ModuleAccess>(..., binary: &M)` instead? Generic type parameters gets expanded during monomorphization and the compiler can optimize the code accordingly. `dyn` requires a vtable lookup at runtime and I don't think we need it here.

Reply:
we'll look into that another time, I don't want to change it now